File: src/parser/enum_parser.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Flow_ast
open Parser_common
open Parser_env
open Token

module Enum (Parse : Parser_common.PARSER) : sig
  val declaration :
    ?leading:Loc.t Comment.t list ->
    const_:bool ->
    env ->
    (Loc.t, Loc.t) Statement.EnumDeclaration.t
end = struct
  open Flow_ast.Statement.EnumDeclaration

  type acc = {
    members: Loc.t member list;
    has_unknown_members: Loc.t option;
    internal_comments: Loc.t Comment.t list;
  }

  type init =
    | NoInit
    | InvalidInit of Loc.t
    | BooleanInit of Loc.t * Loc.t BooleanLiteral.t
    | NumberInit of Loc.t * Loc.t NumberLiteral.t
    | StringInit of Loc.t * Loc.t StringLiteral.t
    | BigIntInit of Loc.t * Loc.t BigIntLiteral.t

  let empty_acc = { members = []; has_unknown_members = None; internal_comments = [] }

  let end_of_member_init env =
    match Peek.token env with
    | T_SEMICOLON
    | T_COMMA
    | T_RCURLY ->
      true
    | _ -> false

  let number_init env loc ~neg ~leading ~kind ~raw =
    let value = Parse.number env kind raw in
    let (value, raw) =
      if neg then
        (-.value, "-" ^ raw)
      else
        (value, raw)
    in
    let trailing = Eat.trailing_comments env in
    if end_of_member_init env then
      NumberInit
        ( loc,
          {
            NumberLiteral.value;
            raw;
            comments = Flow_ast_utils.mk_comments_opt ~leading ~trailing ();
          }
        )
    else
      InvalidInit loc

  let member_init env =
    let loc = Peek.loc env in
    let leading = Peek.comments env in
    match Peek.token env with
    | T_MINUS ->
      Eat.token env;
      (match Peek.token env with
      | T_NUMBER { kind; raw } -> number_init env loc ~neg:true ~leading ~kind ~raw
      | _ -> InvalidInit loc)
    | T_NUMBER { kind; raw } -> number_init env loc ~neg:false ~leading ~kind ~raw
    | T_STRING (loc, value, raw, octal) ->
      if octal then strict_error env Parse_error.StrictOctalLiteral;
      Eat.token env;
      let trailing = Eat.trailing_comments env in
      if end_of_member_init env then
        StringInit
          ( loc,
            {
              StringLiteral.value;
              raw;
              comments = Flow_ast_utils.mk_comments_opt ~leading ~trailing ();
            }
          )
      else
        InvalidInit loc
    | (T_TRUE | T_FALSE) as token ->
      Eat.token env;
      let trailing = Eat.trailing_comments env in
      if end_of_member_init env then
        BooleanInit
          ( loc,
            {
              BooleanLiteral.value = token = T_TRUE;
              comments = Flow_ast_utils.mk_comments_opt ~leading ~trailing ();
            }
          )
      else
        InvalidInit loc
    | T_BIGINT { kind; raw } ->
      let value = Parse.bigint env kind raw in
      let trailing = Eat.trailing_comments env in
      if end_of_member_init env then
        BigIntInit
          ( loc,
            {
              BigIntLiteral.value;
              raw;
              comments = Flow_ast_utils.mk_comments_opt ~leading ~trailing ();
            }
          )
      else
        InvalidInit loc
    | _ ->
      Eat.token env;
      InvalidInit loc

  let member_raw =
    with_loc (fun env ->
        let id = identifier_name env in
        let init =
          match Peek.token env with
          | T_ASSIGN ->
            Expect.token env T_ASSIGN;
            member_init env
          | T_COLON ->
            let (_, { Identifier.name = member_name; _ }) = id in
            error env (Parse_error.EnumInvalidInitializerSeparator { member_name });
            Expect.token env T_COLON;
            member_init env
          | _ -> NoInit
        in
        (id, init)
    )

  let enum_member ~enum_name ~explicit_type acc env =
    let (member_loc, (id, init)) = member_raw env in
    let (_, { Identifier.name = member_name; _ }) = id in
    (* if we parsed an empty name, something has gone wrong and we should abort analysis *)
    if member_name = "" then
      acc
    else
      match init with
      | BooleanInit (loc, value) ->
        let member = BooleanMember (member_loc, { InitializedMember.id; init = (loc, value) }) in
        { acc with members = member :: acc.members }
      | NumberInit (loc, value) ->
        let member = NumberMember (member_loc, { InitializedMember.id; init = (loc, value) }) in
        { acc with members = member :: acc.members }
      | StringInit (loc, value) ->
        let member = StringMember (member_loc, { InitializedMember.id; init = (loc, value) }) in
        { acc with members = member :: acc.members }
      | BigIntInit (loc, value) ->
        let member = BigIntMember (member_loc, { InitializedMember.id; init = (loc, value) }) in
        { acc with members = member :: acc.members }
      | InvalidInit loc ->
        error_at
          env
          (loc, Parse_error.EnumInvalidMemberInitializer { enum_name; explicit_type; member_name });
        acc
      | NoInit ->
        let member = DefaultedMember (member_loc, { DefaultedMember.id }) in
        { acc with members = member :: acc.members }

  let rec enum_members ~enum_name ~explicit_type acc env =
    match Peek.token env with
    | T_RCURLY
    | T_EOF ->
      (List.rev acc.members, acc.has_unknown_members, acc.internal_comments)
    | T_ELLIPSIS ->
      let loc = Peek.loc env in
      (* Internal comments may appear before the ellipsis *)
      let internal_comments = Peek.comments env in
      Eat.token env;
      (match Peek.token env with
      | T_RCURLY
      | T_EOF ->
        ()
      | T_COMMA ->
        Expect.token env T_COMMA;
        let trailing_comma =
          match Peek.token env with
          | T_RCURLY
          | T_EOF ->
            true
          | _ -> false
        in
        error_at env (loc, Parse_error.EnumInvalidEllipsis { trailing_comma })
      | _ -> error_at env (loc, Parse_error.EnumInvalidEllipsis { trailing_comma = false }));
      enum_members
        ~enum_name
        ~explicit_type
        { acc with has_unknown_members = Some loc; internal_comments }
        env
    | _ ->
      let acc = enum_member ~enum_name ~explicit_type acc env in
      (match Peek.token env with
      | T_RCURLY
      | T_EOF ->
        ()
      | T_SEMICOLON ->
        error env Parse_error.EnumInvalidMemberSeparator;
        Expect.token env T_SEMICOLON
      | _ -> Expect.token env T_COMMA);
      enum_members ~enum_name ~explicit_type acc env

  let parse_explicit_type ~enum_name env =
    if Eat.maybe env T_OF then (
      Eat.push_lex_mode env Lex_mode.TYPE;
      let type_loc = Peek.loc env in
      let result =
        match Peek.token env with
        | T_BOOLEAN_TYPE BOOLEAN -> Some (type_loc, Boolean)
        | T_NUMBER_TYPE -> Some (type_loc, Number)
        | T_STRING_TYPE -> Some (type_loc, String)
        | T_SYMBOL_TYPE -> Some (type_loc, Symbol)
        | T_BIGINT_TYPE -> Some (type_loc, BigInt)
        | T_IDENTIFIER { value; _ } ->
          let supplied_type = Some value in
          error env (Parse_error.EnumInvalidExplicitType { enum_name; supplied_type });
          None
        | _ ->
          error env (Parse_error.EnumInvalidExplicitType { enum_name; supplied_type = None });
          None
      in
      Eat.token env;
      Eat.pop_lex_mode env;
      result
    ) else
      None

  let enum_body ~enum_name =
    with_loc (fun env ->
        let explicit_type = parse_explicit_type ~enum_name env in
        let leading =
          if explicit_type <> None then
            Peek.comments env
          else
            []
        in
        Expect.token env T_LCURLY;
        let explicit_type_t = Option.map snd explicit_type in
        let (members, has_unknown_members, internal) =
          enum_members ~enum_name ~explicit_type:explicit_type_t empty_acc env
        in
        let internal = internal @ Peek.comments env in
        Expect.token env T_RCURLY;
        let trailing =
          match Peek.token env with
          | T_EOF
          | T_RCURLY ->
            Eat.trailing_comments env
          | _ when Peek.is_line_terminator env -> Eat.comments_until_next_line env
          | _ -> []
        in
        let comments =
          Flow_ast_utils.mk_comments_with_internal_opt ~leading ~trailing ~internal ()
        in
        { Body.members; explicit_type; has_unknown_members; comments }
    )

  let declaration ?(leading = []) ~const_ env =
    let leading = leading @ Peek.comments env in
    Expect.token env T_ENUM;
    let id = Parse.identifier env in
    let (_, { Identifier.name = enum_name; _ }) = id in
    let body = enum_body ~enum_name env in
    let comments = Flow_ast_utils.mk_comments_opt ~leading () in
    { Statement.EnumDeclaration.id; body; const_; comments }
end

File: rust_port/crates/flow_parser/src/enum_parser.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::ops::Deref;

use dupe::Dupe;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;

use crate::ast::BigIntLiteral;
use crate::ast::BooleanLiteral;
use crate::ast::Comment;
use crate::ast::IdentifierInner;
use crate::ast::NumberLiteral;
use crate::ast::StringLiteral;
use crate::ast::statement;
use crate::ast_utils;
use crate::enum_common::ExplicitType;
use crate::expression_parser;
use crate::loc::LOC_NONE;
use crate::loc::Loc;
use crate::main_parser;
use crate::parse_error::ParseError;
use crate::parser_common;
use crate::parser_common::with_loc;
use crate::parser_env::LexMode;
use crate::parser_env::ParserEnv;
use crate::parser_env::eat;
use crate::parser_env::expect;
use crate::parser_env::peek;
use crate::parser_env::try_parse::Rollback;
use crate::token::BoolOrBoolean;
use crate::token::NumberType;
use crate::token::TokenKind;

struct Acc {
    members: Vec<statement::enum_declaration::Member<Loc>>,
    has_unknown_members: Option<Loc>,
    internal_comments: Vec<Comment<Loc>>,
}

enum Init {
    NoInit,
    InvalidInit(Loc),
    BooleanInit(Loc, BooleanLiteral<Loc>),
    NumberInit(Loc, NumberLiteral<Loc>),
    StringInit(Loc, StringLiteral<Loc>),
    BigIntInit(Loc, BigIntLiteral<Loc>),
}

fn empty_acc() -> Acc {
    Acc {
        members: Vec::new(),
        has_unknown_members: None,
        internal_comments: Vec::new(),
    }
}

fn end_of_member_init(env: &mut ParserEnv) -> bool {
    matches!(
        peek::token(env),
        TokenKind::TSemicolon | TokenKind::TComma | TokenKind::TRcurly
    )
}

fn number_init(
    env: &mut ParserEnv,
    loc: Loc,
    neg: bool,
    leading: Vec<Comment<Loc>>,
    kind: NumberType,
    raw: FlowSmolStr,
) -> Result<Init, Rollback> {
    let value = expression_parser::number(env, kind, &raw)?;
    let (value, raw) = if neg {
        (-value, FlowSmolStr::new(format!("-{}", raw)))
    } else {
        (value, raw)
    };
    let trailing = eat::trailing_comments(env);
    Ok(if end_of_member_init(env) {
        Init::NumberInit(
            loc,
            NumberLiteral {
                value,
                raw,
                comments: ast_utils::mk_comments_opt(Some(leading.into()), Some(trailing.into())),
            },
        )
    } else {
        Init::InvalidInit(loc)
    })
}

fn member_init(env: &mut ParserEnv) -> Result<Init, Rollback> {
    let loc = peek::loc(env).dupe();
    let leading = peek::comments(env);
    match peek::token(env) {
        TokenKind::TMinus => {
            eat::token(env)?;
            match peek::token(env) {
                TokenKind::TNumber { kind, raw } => {
                    let kind = *kind;
                    let raw = raw.to_owned();
                    number_init(env, loc, true, leading, kind, raw)
                }
                _ => Ok(Init::InvalidInit(loc)),
            }
        }
        TokenKind::TNumber { kind, raw } => {
            let kind = *kind;
            let raw = raw.to_owned();
            number_init(env, loc, false, leading, kind, raw)
        }
        TokenKind::TString(str_loc, value, raw, octal) => {
            let str_loc = str_loc.dupe();
            let value = value.to_owned();
            let raw = raw.to_owned();
            if *octal {
                env.strict_error(ParseError::StrictOctalLiteral)?;
            }
            eat::token(env)?;
            let trailing = eat::trailing_comments(env);
            if end_of_member_init(env) {
                Ok(Init::StringInit(
                    str_loc,
                    StringLiteral {
                        value,
                        raw,
                        comments: ast_utils::mk_comments_opt(
                            Some(leading.into()),
                            Some(trailing.into()),
                        ),
                    },
                ))
            } else {
                Ok(Init::InvalidInit(loc))
            }
        }
        TokenKind::TTrue | TokenKind::TFalse => {
            let is_true = peek::token(env) == &TokenKind::TTrue;
            eat::token(env)?;
            let trailing = eat::trailing_comments(env);
            if end_of_member_init(env) {
                Ok(Init::BooleanInit(
                    loc.dupe(),
                    BooleanLiteral {
                        value: is_true,
                        comments: ast_utils::mk_comments_opt(
                            Some(leading.into()),
                            Some(trailing.into()),
                        ),
                    },
                ))
            } else {
                Ok(Init::InvalidInit(loc))
            }
        }
        TokenKind::TBigint { kind, raw } => {
            let kind = *kind;
            let raw = raw.to_owned();
            let value = expression_parser::bigint(env, kind, &raw)?;
            let trailing = eat::trailing_comments(env);
            if end_of_member_init(env) {
                Ok(Init::BigIntInit(
                    loc.dupe(),
                    BigIntLiteral {
                        value,
                        raw,
                        comments: ast_utils::mk_comments_opt(
                            Some(leading.into()),
                            Some(trailing.into()),
                        ),
                    },
                ))
            } else {
                Ok(Init::InvalidInit(loc))
            }
        }
        _ => {
            eat::token(env)?;
            Ok(Init::InvalidInit(loc))
        }
    }
}

fn member_raw(
    env: &mut ParserEnv,
) -> Result<(Loc, (statement::enum_declaration::MemberName<Loc>, Init)), Rollback> {
    with_loc(None, env, |env| {
        let id = match peek::token(env) {
            TokenKind::TString(str_loc, value, raw, octal) => {
                let str_loc = str_loc.dupe();
                let value = value.to_owned();
                let raw = raw.to_owned();
                if *octal {
                    env.strict_error(ParseError::StrictOctalLiteral)?;
                }
                eat::token(env)?;
                statement::enum_declaration::MemberName::StringLiteral(
                    str_loc,
                    StringLiteral {
                        value,
                        raw,
                        comments: None,
                    },
                )
            }
            _ => statement::enum_declaration::MemberName::Identifier(
                parser_common::identifier_name(env)?,
            ),
        };
        let init = match peek::token(env) {
            TokenKind::TAssign => {
                expect::token(env, TokenKind::TAssign)?;
                member_init(env)?
            }
            TokenKind::TColon => {
                let member_name = ast_utils::string_of_enum_member_name(&id).to_owned();
                env.error(ParseError::EnumInvalidInitializerSeparator { member_name })?;
                expect::token(env, TokenKind::TColon)?;
                member_init(env)?
            }
            _ => Init::NoInit,
        };
        Ok((id, init))
    })
}

fn enum_member(
    env: &mut ParserEnv,
    acc: &mut Acc,
    enum_name: &str,
    explicit_type: Option<ExplicitType>,
) -> Result<(), Rollback> {
    let (member_loc, (id, init)) = member_raw(env)?;
    let member_name = ast_utils::string_of_enum_member_name(&id);

    /* if we parsed an empty name, something has gone wrong and we should abort analysis */
    if member_name.is_empty() {
        return Ok(());
    }

    match init {
        Init::BooleanInit(loc, value) => {
            let member = statement::enum_declaration::InitializedMember {
                loc: member_loc,
                id,
                init: (loc, value),
            };
            acc.members
                .push(statement::enum_declaration::Member::BooleanMember(member));
        }
        Init::NumberInit(loc, value) => {
            let member = statement::enum_declaration::InitializedMember {
                loc: member_loc,
                id,
                init: (loc, value),
            };
            acc.members
                .push(statement::enum_declaration::Member::NumberMember(member));
        }
        Init::StringInit(loc, value) => {
            let member = statement::enum_declaration::InitializedMember {
                loc: member_loc,
                id,
                init: (loc, value),
            };
            acc.members
                .push(statement::enum_declaration::Member::StringMember(member));
        }
        Init::BigIntInit(loc, value) => {
            let member = statement::enum_declaration::InitializedMember {
                loc: member_loc,
                id,
                init: (loc, value),
            };
            acc.members
                .push(statement::enum_declaration::Member::BigIntMember(member));
        }
        Init::InvalidInit(loc) => {
            env.error_at(
                loc,
                ParseError::EnumInvalidMemberInitializer {
                    enum_name: enum_name.to_owned(),
                    explicit_type,
                    member_name: member_name.to_owned(),
                },
            )?;
        }
        Init::NoInit => {
            let member = statement::enum_declaration::DefaultedMember {
                loc: member_loc,
                id,
            };
            acc.members
                .push(statement::enum_declaration::Member::DefaultedMember(member));
        }
    }
    Ok(())
}

fn enum_members(
    env: &mut ParserEnv,
    acc: &mut Acc,
    enum_name: &str,
    explicit_type: Option<ExplicitType>,
) -> Result<(), Rollback> {
    loop {
        match peek::token(env) {
            TokenKind::TRcurly | TokenKind::TEof => return Ok(()),
            TokenKind::TEllipsis => {
                let loc = peek::loc(env).dupe();
                // Internal comments may appear before the ellipsis
                let internal_comments = peek::comments(env);
                acc.internal_comments.extend(internal_comments);
                eat::token(env)?;
                match peek::token(env) {
                    TokenKind::TRcurly | TokenKind::TEof => {}
                    TokenKind::TComma => {
                        expect::token(env, TokenKind::TComma)?;
                        let trailing_comma =
                            matches!(peek::token(env), TokenKind::TRcurly | TokenKind::TEof);
                        env.error_at(
                            loc.dupe(),
                            ParseError::EnumInvalidEllipsis { trailing_comma },
                        )?;
                    }
                    _ => {
                        env.error_at(
                            loc.dupe(),
                            ParseError::EnumInvalidEllipsis {
                                trailing_comma: false,
                            },
                        )?;
                    }
                }
                acc.has_unknown_members = Some(loc);
            }
            _ => {
                enum_member(env, acc, enum_name, explicit_type)?;
                match peek::token(env) {
                    TokenKind::TRcurly | TokenKind::TEof => {}
                    TokenKind::TSemicolon => {
                        env.error(ParseError::EnumInvalidMemberSeparator)?;
                        expect::token(env, TokenKind::TSemicolon)?;
                    }
                    _ => {
                        expect::token(env, TokenKind::TComma)?;
                    }
                }
            }
        }
    }
}

fn parse_explicit_type(
    env: &mut ParserEnv,
    enum_name: &str,
) -> Result<Option<(Loc, ExplicitType)>, Rollback> {
    if eat::maybe(env, TokenKind::TOf)? {
        eat::push_lex_mode(env, LexMode::Type);
        let type_loc = peek::loc(env).dupe();
        let result = match peek::token(env) {
            TokenKind::TBooleanType(BoolOrBoolean::Boolean) => {
                Some((type_loc, ExplicitType::Boolean))
            }
            TokenKind::TNumberType => Some((type_loc, ExplicitType::Number)),
            TokenKind::TStringType => Some((type_loc, ExplicitType::String)),
            TokenKind::TSymbolType => Some((type_loc, ExplicitType::Symbol)),
            TokenKind::TBigintType => Some((type_loc, ExplicitType::BigInt)),
            TokenKind::TIdentifier { raw, .. } => {
                let raw = raw.as_str().to_owned();
                env.error(ParseError::EnumInvalidExplicitType {
                    enum_name: enum_name.to_owned(),
                    supplied_type: Some(raw),
                })?;
                None
            }
            _ => {
                env.error(ParseError::EnumInvalidExplicitType {
                    enum_name: enum_name.to_owned(),
                    supplied_type: None,
                })?;
                None
            }
        };
        eat::token(env)?;
        eat::pop_lex_mode(env);
        Ok(result)
    } else {
        Ok(None)
    }
}

fn enum_body(
    env: &mut ParserEnv,
    enum_name: &str,
) -> Result<statement::enum_declaration::Body<Loc>, Rollback> {
    let (loc, mut body) = with_loc(None, env, |env| {
        let explicit_type = parse_explicit_type(env, enum_name)?;
        let leading = if explicit_type.is_some() {
            peek::comments(env)
        } else {
            Vec::new()
        };
        expect::token(env, TokenKind::TLcurly)?;
        let mut acc = empty_acc();
        let explicit_type_kind = explicit_type.as_ref().map(|(_, t)| *t);
        enum_members(env, &mut acc, enum_name, explicit_type_kind)?;
        let mut internal = acc.internal_comments;
        internal.extend(peek::comments(env));
        expect::token(env, TokenKind::TRcurly)?;
        let trailing = match peek::token(env) {
            TokenKind::TEof | TokenKind::TRcurly => eat::trailing_comments(env),
            _ => {
                if peek::is_line_terminator(env) {
                    eat::comments_until_next_line(env)
                } else {
                    Vec::new()
                }
            }
        };
        let comments = ast_utils::mk_comments_with_internal_opt(
            Some(leading.into()),
            Some(trailing.into()),
            Some(internal.into()),
        );
        let members = acc.members;
        let has_unknown_members = acc.has_unknown_members;

        let body = statement::enum_declaration::Body {
            loc: LOC_NONE,
            members: members.into(),
            explicit_type,
            has_unknown_members,
            comments,
        };

        Ok(body)
    })?;
    *body.loc_mut() = loc;
    Ok(body)
}

pub(super) fn declaration(
    leading: Vec<Comment<Loc>>,
    const_: bool,
    env: &mut ParserEnv,
) -> Result<statement::EnumDeclaration<Loc, Loc>, Rollback> {
    let leading = [leading, peek::comments(env)].concat();
    expect::token(env, TokenKind::TEnum)?;
    let id = main_parser::parse_identifier(env, None)?;
    let IdentifierInner {
        loc: _name_loc,
        name: enum_name,
        comments: _,
    } = id.deref();
    let body = enum_body(env, enum_name)?;
    let comments = ast_utils::mk_comments_opt(Some(leading.into()), None);
    Ok(statement::EnumDeclaration {
        id,
        body,
        const_,
        comments,
    })
}
